components/org.wso2.carbon.identity.application.authenticator.kakao/src/main/java/org/wso2/carbon/identity/application/authenticator/kakao/KakaoCustomAuthenticator.java

```diff
@@ -76,21 +123,251 @@ public List<Property> getConfigurationProperties() {
+                String code = getAuthorizationCode(request);
+                String tokenEP = getTokenEndpoint(authenticatorProperties);
+                token = getToken(tokenEP, clientId, clientSecret, code, redirectUri, basicAuthEnabled);
+            }
+
+            Boolean selfContainedTokenEnabled = Boolean.parseBoolean(authenticatorProperties
+                    .get(Oauth2GenericAuthenticatorConstants.SELF_CONTAINED_TOKEN_ENABLED));
+            String userInfo = getUserInfo(selfContainedTokenEnabled, token, authenticatorProperties);
+            buildClaims(context, userInfo);
+        } catch (ApplicationAuthenticatorException | MisconfigurationException e) {
+            String errorMessage = "Error while processing authentication response.";
+            throw new AuthenticationFailedException(errorMessage, e);
+        }
+    }
+
+    private boolean isTrustedTokenIssuer(AuthenticationContext context) {
+
+        ExternalIdPConfig externalIdPConfig = context.getExternalIdP();
+        if (externalIdPConfig == null) {
+            return false;
+        }
+
+        IdentityProvider externalIdentityProvider = externalIdPConfig.getIdentityProvider();
+        if (externalIdentityProvider == null) {
+            return false;
+        }
+
+        IdentityProviderProperty[] identityProviderProperties = externalIdentityProvider.getIdpProperties();
+        for (IdentityProviderProperty identityProviderProperty : identityProviderProperties) {
+            if (IdPManagementConstants.IS_TRUSTED_TOKEN_ISSUER.equals(identityProviderProperty.getName())) {
+                return Boolean.parseBoolean(identityProviderProperty.getValue());
+            }
+        }
+
+        return false;
+    }
+
+    private boolean isNativeSDKBasedFederationCall(HttpServletRequest request) {
+
+        return request.getParameter(ACCESS_TOKEN_PARAM) != null && request.getParameter(ID_TOKEN_PARAM) != null;
+    }
+
+    private AdditionalData getAdditionalData(AuthenticationContext context, boolean isNativeSDKBasedFederationCall) {
+
+        AdditionalData additionalData = new AdditionalData();
+
+        if (isNativeSDKBasedFederationCall) {
+            Map<String, String> additionalAuthenticationParams = new HashMap<>();
+            additionalData.setAdditionalAuthenticationParams(additionalAuthenticationParams);
+        } else {
+            Map<String, String> additionalAuthenticationParams = new HashMap<>();
+            additionalAuthenticationParams.put(REDIRECT_URL, context.getProperty(REDIRECT_URL).toString());
+            additionalAuthenticationParams.put(STATE, context.getProperty(STATE).toString());
+            additionalData.setAdditionalAuthenticationParams(additionalAuthenticationParams);
+        }
+        return additionalData;
+    }
+
+    private void validateJWTToken(AuthenticationContext context, String idToken) throws AuthenticationFailedException,
```

Comment:
Haven't we done these validations with the existing authenticator implementation? Why is it specifically needed for API based approach?

Reply:
In the OAuth2GenericAuthenticator from which the KakaoCustomAuthenticator is extended from, we don't read or validate the id token. We work with the access token there.

https://github.com/wso2-extensions/identity-outbound-auth-oauth2/blob/e9920ac286b77b502d3d60ed5c2a00ccbf5d0e0e/components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java#L197